[SOL-1953] Update Commerce Configuration

marjev opened the pull request:
@mjfrey Here's the PR for updating Commerce Configuration.

I connected Commerce Configuration with the Site model so that someone can add a Site and connect it with Commerce. By default, Site is not provided (read: fresh devstack and full-stack environments) and in that case, ECOMMERCE_PUBLIC_URL_ROOT is used. If a site is provided and the current configuration has a site associated with it, that site's domain name will be used as a root.

So, we have a URL of the following form: {ROOT}{RECEIPT_URL}{ORDER_NUMBER}.

ROOT is either ECOMMERCE_PUBLIC_URL_ROOT or Site domain name. I assumed Site domain name will contain https:// prefix. If not, please let me know. I'm not sure how WL Sites are configured exactly.
RECEIPT_URL should be a relative path to the receipt page that's appended to the ROOT.

Let me know if this is the right way to do it and if you have any suggestions/questions/concerns.

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=13299&repo=edx%2Fedx-platform

mjfrey commented on lms/djangoapps/student_account/views.py:
```diff
@@ -331,6 +331,8 @@ def get_user_orders(user):
         commerce_configuration, user, 'orders', api=api, querystring=user_query, cache_key=cache_key
     )
 
+    commerce_root = commerce_configuration.site or settings.ECOMMERCE_PUBLIC_URL_ROOT
```
we need a way to use either otto or lms for the root.  So we can switch back and forth between the LMS hosted receipt page and the otto one.

mjfrey commented:
@mattdrayer please take a look.

mjfrey commented on lms/djangoapps/commerce/models.py:
```diff
@@ -35,9 +36,16 @@ class Meta(object):
+        default='/checkout/receipt/?orderNum=',
         help_text=_('Path to order receipt page.')
     )
+    site = models.ForeignKey(
```
Leave commerce configuration the way it is.  Create new fields in the site config model.

marjev replied:
We'll need this ForeignKey so that we may define switch back and forth between the two sites (LMS and Otto). I can, however, move the receipt page there so that we may define the relative path for each site.

marjev commented:
@mjfrey Please take a look at the 2nd commit. So, the idea is that if no Site is setup, the system will use LMS receipt page by default. I've added a method to the CommerceConfiguration model so that the view can see if the receipt page is hosted by LMS or some other Site.

Other sites (e.g. Otto) will have to be added to the Site model and configured using SiteConfiguration model. CommerceConfiguration will have to be connected to the Site if it's necessary to use the receipt page of that Site. So, for that to work, I moved the receipt_page field from CommerceConfiguration to the SiteConfiguration.

mattdrayer commented on lms/djangoapps/commerce/migrations/0005_auto_20160825_0544.py:
```diff
@@ -0,0 +1,25 @@
+        migrations.AddField(
+            model_name='commerceconfiguration',
+            name='site',
+            field=models.ForeignKey(related_name='commerce_configuration', on_delete=django.db.models.deletion.SET_NULL, blank=True, to='sites.Site', null=True),
```
I would keep the related name as the default 'commerceconfiguration' here, unless there is a technical reason to change it.

mattdrayer commented on lms/djangoapps/commerce/models.py:
```diff
@@ -33,15 +38,36 @@ class Meta(object):
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
+        related_name='commerce_configuration'
```
I guess this is the right spot to focus on in terms of related_name -- I would prefer the RN to remain 'commerceconfiguration' if possible.

mattdrayer commented on lms/djangoapps/commerce/models.py:
```diff
@@ -33,15 +38,36 @@ class Meta(object):
+                site_configuration = SiteConfiguration.objects.get(site=site)
+                return '{site_domain}{receipt_page}'.format(
+                    site_domain=site.domain,
+                    receipt_page=site_configuration.receipt_page
```
can you simply do site.site_configuration.receipt_page here?

mattdrayer commented on lms/djangoapps/student_account/views.py:
```diff
@@ -309,7 +309,7 @@ def _external_auth_intercept(request, mode):
 
 
-def get_user_orders(user):
+def get_user_orders(user, site):
```
It appears that `site` is not actually being used here

marjev commented:
@mattdrayer Thank you for reviewing this PR. I've addressed your comments. Could you take another look?

mattdrayer commented on lms/djangoapps/student_account/views.py:
```diff
@@ -347,7 +349,10 @@ def get_user_orders(user):
                                         date_placed.replace(tzinfo=pytz.UTC), 'SHORT_DATE'
                                     ),
-                                    'receipt_url': commerce_configuration.receipt_page + order['number']
+                                    'receipt_url': '{receipt_page_url}{order_number}'.format(
```
Does it make sense to pass the order number to the receipt page URL and have that be responsible for the full construction?  It seems like this is a critical piece of data for the URL to actually be useful.

mattdrayer commented on openedx/core/djangoapps/site_configuration/models.py:
```diff
@@ -32,6 +33,11 @@ class SiteConfiguration(models.Model):
         blank=True,
         load_kwargs={'object_pairs_hook': collections.OrderedDict}
     )
+    receipt_page = models.CharField(
```
I think a better name for this field is "receipt_page_path"

mattdrayer commented on lms/djangoapps/commerce/models.py:
```diff
@@ -33,15 +37,35 @@ class Meta(object):
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
+        related_name='commerceconfiguration'
```
What is the default related_name here?  If it is "commerceconfiguration" then you can drop this attribute altogether.

marjev commented:
@mattdrayer Comments have been addressed. Please take a look at 19f6a5a4cfc900163b0caacd1cc49e5fcbf25114.

@vkaracic Could you take a look as well?

mattdrayer commented on openedx/core/djangoapps/site_configuration/models.py:
```diff
@@ -32,6 +33,11 @@ class SiteConfiguration(models.Model):
     )
+    receipt_page_url = models.CharField(
+        max_length=255,
+        default='/commerce/checkout/receipt/?orderNum=',
```
This default value is a duplicate of the failover value used [here](https://github.com/edx/edx-platform/pull/13299/files#diff-6bfaaa3703d73979b336ebbd7a656181R67) -- it probably makes sense to move it into a constant or configuration setting

mattdrayer commented:
Almost there @marjev -- if you can address my last comment and get the checks to pass, it's a 👍  for me.

marjev commented:
jenkins run bokchoy

vkaracic commented on lms/djangoapps/commerce/models.py:
```diff
@@ -33,15 +38,38 @@ class Meta(object):
+        """
+        Return absolute receipt page URL.
+
+        Returns:
```
Nit: args missing.

vkaracic commented:
Is a test needed? Just to cover the new method.
The code LGTM.

marjev commented:
jenkins run all

marjev commented:
@mjfrey can you please take a final look before this is merged?

mjfrey commented:
👍
